odl/space/pspace.py

```diff
@@ -381,6 +384,45 @@ def dtype(self):
         else:
             raise AttributeError("`dtype`'s of subspaces not equal")
 
+    @property
+    def real_space(self):
+        """Variant of this space with real dtype."""
+        return ProductSpace(*[space.real_space for space in self.spaces])
+
+    @property
+    def complex_space(self):
+        """The space corresponding to this space's complex dtype."""
+        return ProductSpace(*[space.complex_space for space in self.spaces])
+
+    def astype(self, dtype):
+        """Return a copy of this space with new ``dtype``.
+
+        Parameters
+        ----------
+        dtype :
+            Scalar data type of the returned space. Can be provided
+            in any way the `numpy.dtype` constructor understands, e.g.
+            as built-in type or as a string. Data types with non-trivial
+            shapes are not allowed.
+
+        Returns
+        -------
+        newspace : `ProductSpace`
+            Version of this space with given data type.
+        """
+        if dtype is None:
+            # Need to filter this out since Numpy iterprets it as 'float'
+            raise ValueError('`None` is not a valid data type')
+
+        dtype = np.dtype(dtype)
+        current_dtype = getattr(self, 'dtype', None)
+
+        if dtype == current_dtype:
+            return self
```

Comment:
There's a bug here since 
```python
>>> np.dtype(float) == None
True
```
which means that `self` is returned if `dtype == np.dtype(float)` and no dtype of the current space can be determined. That's why I suggested `object`, not `None`, as `getattr` fallback (which assumes we will never support `object`).

Reply:
If this is never exposed externally, prefer e.g. `-1` as a fallback value since we know for sure we'll never support that.

Reply:
-1 raises an error that we would need to catch. Alternatively we could use `np.void` which is about the only thing that compares to `False` with `np.dtype(object)`.